File: task1/problem_h.cpp
#include <iostream>
#include <vector>


struct Node {
    Node(int data_to_store) {
        data = data_to_store;
        next = nullptr;
    }

    int data;
    Node *next;
};

class Queue {
public:
    void push_back(int data) {
        Node *new_node = new Node(data);
        if (len == 0) {
            begin = end = new_node;
        } else {
            end->next = new_node;
            end = new_node;
        }
        len++;
        if (min_node == nullptr || data <= min) {
            min_node = new_node;
            min = data;
        }
    }

    void __find_min() {
        Node *node = begin;
        min = node->data;
        min_node = node;
        while (node->next != nullptr) {
            node = node->next;
            if (node->data <= min) {
                min = node->data;
                min_node = node;
            }
        }
    }

    int pop_first() {
        Node* node_to_return = begin;
        int data_to_return = node_to_return->data;
        if (len == 1) {
            begin = end = min_node = nullptr;
            min = 0;
        } else {
            begin = begin->next;
            if (node_to_return == min_node) {
                __find_min();
            }
        }
        len--;
        delete(node_to_return);
        return data_to_return;
    }

    int find_min() {
        return min;
    }

    ~Queue() {
        while (len > 0) {
            pop_first();
        }
    }
private:
    int len{0};
    Node *begin{nullptr};
    Node *end{nullptr};
    int min{0};
    Node *min_node{nullptr};
};


int main() {
    int n, k;
    std::cin >> n >> k;
    std::vector<int> sequence(n);
    int value;
    for (int i = 0; i < n; ++i) {
        std::cin >> value;
        sequence[i] = value;
    }

    Queue window = Queue();
    for (int i = 0; i < k; ++i) {
        window.push_back(sequence[i]);
    }
    std::cout << window.find_min() << std::endl;

    for (int i = k; i < n; ++i) {
        window.pop_first();
        window.push_back(sequence[i]);
        std::cout << window.find_min() << std::endl;
    }
    return 0;
}

File: task1/problem_h.py
#
# This solution doesn't reach good results, so look at the c++ implementation
#
n, k = [int(x) for x in input().strip().split()]
sequence = [int(x) for x in input().strip().split()]


class Queue:
    class Node:
        def __init__(self, data):
            self.data = data
            self.next = None

    def __init__(self):
        self.len = 0
        self.begin = self.end = None
        self.min = None
        self.last_min = None
        
    def push_back(self, data):
        new_node = self.Node(data)
        if self.len == 0:
            self.begin = self.end = new_node
        else:
            self.end.next = new_node
            self.end = new_node
        self.len += 1
        if self.last_min is None or data <= self.min:
            self.last_min = new_node
            self.min = data
            
    def __linear_find_min(self):
#         assert(self.len > 0)
        node = self.begin
        self.min = node.data
        self.last_min = self.begin
        while node.next is not None:
            node = node.next
            if node.data <= self.min:
                self.last_min = node
                self.min = node.data
            
    def pop_first(self):
#         assert(self.len > 0)
        node_to_return = self.begin
        if self.len == 1:
            self.begin = self.end = None
            self.last_min = self.min = None
        else:
            self.begin = self.begin.next
            if node_to_return == self.last_min:
                self.__linear_find_min()
        self.len -= 1
        return node_to_return.data
    
    def find_min(self):
        assert(self.len > 0)
        return self.min

    
window = Queue()
for i in range(k):
    window.push_back(sequence[i])
print(window.find_min())

for i in range(k, n):
    window.pop_first()
    window.push_back(sequence[i])
    print(window.find_min())
